src/java.base/share/classes/java/lang/Runtime.java

```diff
@@ -48,7 +48,51 @@
  * <p>
  * An application cannot create its own instance of this class.
  *
+ * <h2><a id="shutdown">Shutdown Sequence</a></h2>
+ * <p>
+ * The Java Virtual Machine initiates the <i>shutdown sequence</i> in response
+ * to one of several events:
+ * <ol>
+ * <li>when the number of non-daemon threads drops to zero for the first time;</li>
+ * <li>when the {@link #exit Runtime.exit} or {@link System#exit System.exit} method is called; or</li>
+ * <li>when some external event occurs, such as an interrupt or a signal is received from
+ * the operating system.</li>
+ * </ol>
+ *
+ * <p>At the beginning of the shutdown sequence, the set of registered
+ * {@linkplain #addShutdownHook shutdown hooks}
+ * is established atomically. After this point, registration and de-registration
+ * of shutdown hooks with
+ * {@link #addShutdownHook addShutdownHook}
+ * and
+ * {@link #removeShutdownHook removeShutdownHook}
+ * is prohibited. The shutdown hooks are then started in some unspecified order,
+ * and they are allowed to run concurrently. Any already-existing threads and any
+ * newly started threads also continue to run concurrently during the shutdown sequence.
+ * <p>
+ * When all shutdown hooks have completed, the Java Virtual Machine terminates as described
+ * below. If for some reason one of the shutdown hooks does not complete, if for example
+ * {@link #exit exit} is called, this will prevent the shutdown sequence from completing.
+ * In this case other threads (including shutdown hooks) continue to execute and can
+ * terminate the JVM via the {@link #halt halt} method.
+ * <h2><a id="termination">Virtual Machine Termination</a></h2>
+ * When the JVM terminates, all threads (daemon, non-daemon, and shutdown hooks) are immediately
```

Comment:
"daemon, non-daemon, and shutdown hooks" reads like there are 3 kinds of threads. I would be tempted to drop "shutdown hooks" from this list or else phrase it as "all shutdown hooks and all other threads".

Reply:
Alex and I discussed this and we felt that it was important to call out all three "kinds" of threads explicitly. Earlier versions of the text implied that no non-daemon threads were running at the time the VM terminated, which might not be true. Earlier text also implied that shutdown hooks had terminated before the VM terminated, but this isn't necessarily true if somebody calls halt(). Yes, a shutdown hook has a daemon/non-daemon bit, but effectively it's ignored, and a shutdown hook is treated so specially that it's effectively a different kind of thread.

Reply:
I understand but as currently worded the reader may wonder what Thread::isDaemon might return when it invoked on Thread that is a shutdown hook. I think the best I can suggest is to replace "daemon, non-daemon, and shutdown hooks" with "shutdown hooks and other daemon and non-daemon threads".

Reply:
OK I used this wording.